hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/main/java/org/apache/hadoop/mapreduce/v2/app/launcher/ContainerLauncherImpl.java

```diff
@@ -385,6 +385,20 @@ public void run() {
       // TODO: Do it only once per NodeManager.
       ContainerId containerID = event.getContainerID();
 
+      // If the container failed to launch earlier (due to dead node for example),
+      // it has been marked as FAILED and removed from containers during
+      // CONTAINER_REMOTE_LAUNCH event handling.
+      // Skip kill() such container during CONTAINER_REMOTE_CLEANUP as
+      // it is not necessary and could cost 15 minutes delay if the node is dead.
+      if (event.getType() == EventType.CONTAINER_REMOTE_CLEANUP &&
+          !containers.containsKey(containerID)) {
+        LOG.info("Skip cleanup of already-removed container " + containerID);
```

Comment:
Use logger {}

Reply:
As we are at it, fix this in line 382 too.

Reply:
Ack, Will make this change.